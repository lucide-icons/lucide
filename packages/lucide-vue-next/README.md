# Lucide Vue Next

Implementation of the lucide icon library for Vue 3 applications.

> What is lucide? Read it [here](https://github.com/lucide-icons/lucide#what-is-lucide).

> :warning: This version of lucide is for Vue 3, For Vue 2 got to [lucide-vue](https://github.com/lucide-icons/lucide/tree/main/packages/lucide-vue#lucide-vue)

## Installation

```sh
yarn add lucide-vue-next
```

or

```sh
npm install lucide-vue-next
```

## How to use

It's build with ESmodules so it's completely tree-shakable.
Each icon can be imported as a vue component.

### Example

You can pass additional props to adjust the icon.

```vue
<template>
  <Camera color="red" :size="32" />
</template>

<script>
// Returns Vue component
import { Camera } from 'lucide-vue-next';

export default {
  name: 'My Component',
  components: { Camera }
};
</script>
```

### Props

|  name                   |   type    |  default     |
| ----------------------- | --------- | ------------ |
| `size`                  | *number*  | 24           |
| `color`                 | *string*  | currentColor |
| `stroke-width`          | *number*  | 2            |
| `absolute-stroke-width` | *boolean* | false        |
| `default-class`         | *string*  | lucide-icon  |

### Custom props

You can also pass custom props that will be added in the svg as attributes.

```vue
<template>
  <Camera fill="red" />
</template>
```

### One generic icon component

It is possible to create one generic icon component to load icons.

> :warning: Example below importing all EsModules, caution using this example, not recommended when you using bundlers, your application build size will grow strongly.

#### Icon Component Example

```vue
<template>
  <component :is="icon" />
</template>

<script>
import * as icons from 'lucide-vue-next';

export default {
  props: {
    name: {
      type: String,
      required: true
    }
  },
  computed: {
    icon() {
      return icons[this.name];
    }
  }
};
</script>
```

##### Then you can use it like this

```vue
<template>
  <div id="app">
    <Icon name="Airplay" />
  </div>
</template>
```
