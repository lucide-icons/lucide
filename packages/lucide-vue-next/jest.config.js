module.exports = {
  verbose: true,
  roots: ['<rootDir>/src/', '<rootDir>/tests/'],
  moduleFileExtensions: ['js'],
  transform: {
    '^.+\\.js$': 'babel-jest',
    '^.+\\.vue$': 'vue-jest',
  },
  transformIgnorePatterns: [`/node_modules`],
  snapshotSerializers: ['jest-serializer-vue'],
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1',
  },
};
