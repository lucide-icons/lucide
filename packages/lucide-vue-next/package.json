{
  "name": "lucide-vue-next",
  "version": "0.15.17-beta.0",
  "author": "Eric Fennis",
  "description": "Lucide Vue 3 Package",
  "license": "ISC",
  "homepage": "https://lucide.dev",
  "bugs": "https://github.com/lucide-icons/lucide/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/lucide-icons/lucide.git",
    "directory": "packages/lucide-vue-next"
  },
  "amdName": "lucide-vue-next",
  "source": "build/lucide-vue-next.js",
  "main": "dist/cjs/lucide-vue-next.js",
  "main:umd": "dist/umd/lucide-vue-next.js",
  "module": "dist/esm/lucide-vue-next.js",
  "unpkg": "dist/umd/lucide-vue-next.min.js",
  "typings": "dist/lucide-vue-next.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "yarn clean && yarn build:icons && yarn build:es && yarn build:types && yarn build:bundles",
    "clean": "rm -rf dist && rm -rf ./src/icons/*.js",
    "build:icons": "yarn --cwd ../../ build:icons --output=../packages/lucide-vue-next/src --templateSrc=../packages/lucide-vue-next/scripts/exportTemplate",
    "build:es": "babel src -d dist/esm",
    "build:types": "yarn --cwd ../../ babel-node packages/lucide-vue-next/scripts/buildTypes.js",
    "build:bundles": "yarn --cwd ../../ rollup -c packages/lucide-vue-next/rollup.config.js",
    "test": "jest",
    "test:watch": "jest --watchAll"
  },
  "devDependencies": {
    "@vue/compiler-sfc": "^3.0.0",
    "@vue/test-utils": "^2.0.0-rc.6",
    "jest-serializer-vue": "^2.0.2",
    "vue-jest": "^5.0.0-alpha.7",
    "vue": "3.0.6"
  },
  "peerDependencies": {
    "vue": "^3.0.0"
  }
}
