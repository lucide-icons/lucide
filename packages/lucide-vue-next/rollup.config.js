import plugins from '../../rollup.plugins';
import pkg from './package.json';

const packageName = 'LucideVueNext';
const outputFileName = 'lucide-vue-next';
const rootDir = 'packages/lucide-vue-next'; // It runs from the root
const outputDir = `${rootDir}/dist`;
const inputs = [`${rootDir}/src/lucide-vue-next.js`];
const bundles = [
  {
    format: 'umd',
    inputs,
    outputDir,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
    outputDir,
  },
  {
    format: 'cjs',
    inputs,
    outputDir,
  },
];

const configs = bundles
  .map(({ inputs, outputDir, format, minify }) =>
    inputs.map(input => ({
      input,
      plugins: plugins(pkg, minify),
      external: ['vue'],
      output: {
        name: packageName,
        file: `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.js`,
        format,
        sourcemap: true,
        globals: {
          vue: 'vue',
        },
      },
    })),
  )
  .flat();

export default configs;
