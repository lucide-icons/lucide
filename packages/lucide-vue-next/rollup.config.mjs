import plugins, { replace } from '@lucide/rollup-plugins';
import pkg from './package.json' assert { type: 'json' };

const packageName = 'LucideVueNext';
const outputFileName = 'lucide-vue-next';
const outputDir = 'dist';
const inputs = ['src/lucide-vue-next.ts'];
const bundles = [
  {
    format: 'umd',
    inputs,
    outputDir,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
    outputDir,
  },
  {
    format: 'cjs',
    inputs,
    outputDir,
    aliasesSupport: true
  },
  {
    format: 'esm',
    inputs,
    outputDir,
    preserveModules: true,
    aliasesSupport: true
  },
];

const configs = bundles
  .map(({ inputs, outputDir, format, minify, preserveModules, aliasesSupport }) =>
    inputs.map(input => ({
      input,
      plugins: [
        // This for aliases, only for esm
        ...(
          !aliasesSupport ? [
            replace({
              "export * from './aliases';": '',
              "export * as icons from './icons';": '',
              delimiters: ['', ''],
              preventAssignment: false,
            }),
          ] : []
        ),
        ...plugins(pkg, minify)
      ],
      external: ['vue'],
      output: {
        name: packageName,
        ...(preserveModules
          ? {
              dir: `${outputDir}/${format}`,
            }
          : {
              file: `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.js`,
            }),
        format,
        preserveModules,
        sourcemap: true,
        globals: {
          vue: 'vue',
        },
      },
    })),
  )
  .flat();

export default configs;
