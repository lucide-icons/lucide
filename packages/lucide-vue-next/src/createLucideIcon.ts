import { h } from 'vue';
import type { SVGAttributes, FunctionalComponent } from 'vue';
import defaultAttributes from './defaultAttributes';

// Create interface extending SVGAttributes
export interface SVGProps extends Partial<SVGAttributes> {
  size?: 24 | number
}


type IconNode = [elementName: string, attrs: Record<string, string>][]

/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string) => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

const createLucideIcon = (iconName: string, iconNode: IconNode): FunctionalComponent<SVGProps> => (
  { size, color, ...props }, // props
  { attrs, slots } // context
  ) => {
  return h(
    'svg',
    {
      ...defaultAttributes,
      width: size || defaultAttributes.width,
      height: size || defaultAttributes.height,
      stroke: color || defaultAttributes.stroke,
      ...attrs,
      class: ['lucide', `lucide-${toKebabCase(iconName)}`, attrs?.class || ''],
      ...props,
    },
    [
      ...iconNode.map(child => h(...child)),
      ...(slots.default ? [slots.default()] : [])
    ],
  );
};

export default createLucideIcon;
