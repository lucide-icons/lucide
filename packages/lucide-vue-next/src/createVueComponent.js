import { h } from 'vue';
import defaultAttributes from './defaultAttributes';

const createVueComponent = (iconName, iconNode) => (props, context) =>
  h(
    'svg',
    {
      ...defaultAttributes,
      ...{
        width: props.size || defaultAttributes.width,
        height: props.size || defaultAttributes.height,
      },
      ...context.attrs,
      ...props,
    },
    iconNode.map(child => h(...child)),
  );

export default createVueComponent;
