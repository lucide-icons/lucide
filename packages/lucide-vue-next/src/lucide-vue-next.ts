export * from './icons';
export * as icons from './icons';
export * from './aliases';

export { default as createLucideIcon } from './createLucideIcon';
export { default as useIconComponent } from './useIconComponent';
