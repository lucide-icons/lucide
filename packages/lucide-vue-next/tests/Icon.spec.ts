import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/vue';

import { airVent } from './testIconNodes';
import { Icon } from '../src/lucide-vue-next';

describe('Using Icon Component', () => {
  it('should render icon based on a iconNode', async () => {
    const { container } = render(Icon, {
      props: {
        iconNode: airVent,
        size: 48,
        color: 'red',
        absoluteStrokeWidth: true,
      },
    });

    expect(container.firstChild).toBeDefined();
  });

  it('should render icon and match snapshot', async () => {
    const { container } = render(Icon, {
      props: {
        iconNode: airVent,
        size: 48,
        color: 'red',
        absoluteStrokeWidth: true,
      },
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
