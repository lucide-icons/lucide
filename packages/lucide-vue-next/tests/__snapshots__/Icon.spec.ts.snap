// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Using Icon Component > should render icon and match snapshot 1`] = `
<svg
  class="lucide lucide-icon"
  fill="none"
  height="48"
  stroke="red"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="1"
  viewBox="0 0 24 24"
  width="48"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6 12H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v5a2 2 0 0 1-2 2h-2"
  />
  <path
    d="M6 8h12"
  />
  <path
    d="M18.3 17.7a2.5 2.5 0 0 1-3.16 3.83 2.53 2.53 0 0 1-1.14-2V12"
  />
  <path
    d="M6.6 15.6A2 2 0 1 0 10 17v-5"
  />
</svg>
`;
